<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ THmisc</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">THmisc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_cluster_infos.html">add_cluster_infos()</a></code> </p>
        </td>
        <td><p>Adds a 'cluster' id to a mutation data frame</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_mutation_types.html">annotate_mutation_types()</a></code> </p>
        </td>
        <td><p>Create detailed mutation annotation.</p></td>
      </tr><tr><td>
          <p><code><a href="annotation_from_barcode.html">annotation_from_barcode()</a></code> </p>
        </td>
        <td><p>Get sample annotations from sample barcodes.</p></td>
      </tr><tr><td>
          <p><code><a href="annotation_from_barcode_epicc.html">annotation_from_barcode_epicc()</a></code> </p>
        </td>
        <td><p>EPICC Project: Get sample annotations from sample barcodes.</p></td>
      </tr><tr><td>
          <p><code><a href="annotation_from_barcode_tcga.html">annotation_from_barcode_tcga()</a></code> </p>
        </td>
        <td><p>TCGA project: Create a annotation for a vector of TCGA barcodes.</p></td>
      </tr><tr><td>
          <p><code><a href="calc_avg_cn_per_bin.html">calc_avg_cn_per_bin()</a></code> </p>
        </td>
        <td><p>Calculate average copy-number values in bins.</p></td>
      </tr><tr><td>
          <p><code><a href="calc_cn_summary_stats.html">calc_cn_summary_stats()</a></code> </p>
        </td>
        <td><p>Calculate copy-number alteration summary stats</p></td>
      </tr><tr><td>
          <p><code><a href="calc_pga_stats.html">calc_pga_stats()</a></code> </p>
        </td>
        <td><p>Calculate PGA stats</p></td>
      </tr><tr><td>
          <p><code><a href="drop_multiallelic.html">drop_multiallelic()</a></code> </p>
        </td>
        <td><p>Function to drop multiallelic sites from VCFs</p></td>
      </tr><tr><td>
          <p><code><a href="gene_lists.html">gene_lists</a></code> </p>
        </td>
        <td><p>List of genes for various analysis</p></td>
      </tr><tr><td>
          <p><code><a href="get_ab_alleles.html">get_ab_alleles()</a></code> </p>
        </td>
        <td><p>Get A&amp;B allele numbers from a list of CN data.</p></td>
      </tr><tr><td>
          <p><code><a href="get_cnas.html">get_cnas()</a></code> </p>
        </td>
        <td><p>Get CN annotations from a list of cn data.</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_pos.html">get_gene_pos()</a></code> </p>
        </td>
        <td><p>Function to look-up (average) gene positions.</p></td>
      </tr><tr><td>
          <p><code><a href="get_mutation_df.html">get_mutation_df()</a></code> </p>
        </td>
        <td><p>Get a data frame of mutation from various input formats</p></td>
      </tr><tr><td>
          <p><code><a href="get_mutation_type.html">get_mutation_type()</a></code> </p>
        </td>
        <td><p>Get the mutation type of mutations</p></td>
      </tr><tr><td>
          <p><code><a href="get_transition_type.html">get_transition_type()</a></code> </p>
        </td>
        <td><p>Get the transition type of mutations</p></td>
      </tr><tr><td>
          <p><code><a href="indentify_vcf_type.html">indentify_vcf_type()</a></code> </p>
        </td>
        <td><p>Function to identify which tool created a given vcf file</p></td>
      </tr><tr><td>
          <p><code><a href="insert_ccf.html">insert_ccf()</a></code> </p>
        </td>
        <td><p>Adds an additional CCF entry to a 'CollapsedVCF' object.</p></td>
      </tr><tr><td>
          <p><code><a href="insert_cnas.html">insert_cnas()</a></code> </p>
        </td>
        <td><p>Add an additional CNA and AB entries to a 'CollapsedVCF' object.</p></td>
      </tr><tr><td>
          <p><code><a href="insert_vaf.html">insert_vaf()</a></code> </p>
        </td>
        <td><p>Adds an additional VAF entry to a 'CollapsedVCF' object.</p></td>
      </tr><tr><td>
          <p><code><a href="lift_data_wrapper.html">lift_data_wrapper()</a></code> </p>
        </td>
        <td><p>Internal function used to lift mutation data between genome versions.</p></td>
      </tr><tr><td>
          <p><code><a href="lift_hg19_to_hg38.html">lift_hg19_to_hg38()</a></code> </p>
        </td>
        <td><p>Function to lift mutations or objects between genome versions</p></td>
      </tr><tr><td>
          <p><code><a href="lift_hg38_to_hg19.html">lift_hg38_to_hg19()</a></code> </p>
        </td>
        <td><p>Function to lift mutations or objects between genome versions</p></td>
      </tr><tr><td>
          <p><code><a href="load_sequenza_purity_files.html">load_sequenza_purity_files()</a></code> </p>
        </td>
        <td><p>Convience function to load Sequenza fit files.</p></td>
      </tr><tr><td>
          <p><code><a href="load_sequenza_segment_files.html">load_sequenza_segment_files()</a></code> </p>
        </td>
        <td><p>Convience function to load Sequenza segment files.</p></td>
      </tr><tr><td>
          <p><code><a href="load_vcf_file.html">load_vcf_file()</a></code> </p>
        </td>
        <td><p>Wrapper function to load arbitrary VCF files.</p></td>
      </tr><tr><td>
          <p><code><a href="long_to_short_peptide_names.html">long_to_short_peptide_names()</a></code> </p>
        </td>
        <td><p>Convert long peptide names to short ones.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cn_altered_per_patient.html">plot_cn_altered_per_patient()</a></code> </p>
        </td>
        <td><p>Plot per-case copy-number alteration summary stats</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cn_freq.html">plot_cn_freq()</a></code> </p>
        </td>
        <td><p>Plot copy-number alteration summary stats across cases</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cn_states_across_samples.html">plot_cn_states_across_samples()</a></code> </p>
        </td>
        <td><p>Plot CN states of somatic sides across samples.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_driver_vaf_heatmap.html">plot_driver_vaf_heatmap()</a></code> <code><a href="plot_driver_vaf_heatmap.html">save_driver_vaf_heatmap()</a></code> </p>
        </td>
        <td><p>Plot a heatmap showing VAF of specific mutations across samples.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_gene_mut_heatmap.html">plot_gene_mut_heatmap()</a></code> </p>
        </td>
        <td><p>Create a plot showing if a gene is mutated in a given sample</p></td>
      </tr><tr><td>
          <p><code><a href="plot_inter_mutation_distance.html">plot_inter_mutation_distance()</a></code> </p>
        </td>
        <td><p>Function to plot intermutation distances</p></td>
      </tr><tr><td>
          <p><code><a href="plot_mutation_heatmap.html">plot_mutation_heatmap()</a></code> </p>
        </td>
        <td><p>Function to plot mutation data as a heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plot_mutation_histogram.html">plot_mutation_histogram()</a></code> </p>
        </td>
        <td><p>Function to plot mutation data as a histogram</p></td>
      </tr><tr><td>
          <p><code><a href="plot_pga_stats.html">plot_pga_stats()</a></code> </p>
        </td>
        <td><p>Plot PGA stats</p></td>
      </tr><tr><td>
          <p><code><a href="plot_vaf_distribution.html">plot_vaf_distribution()</a></code> </p>
        </td>
        <td><p>Plot a histogram of mutations contained in a VCF across CN states.</p></td>
      </tr><tr><td>
          <p><code><a href="pretty_violin.html">pretty_violin()</a></code> </p>
        </td>
        <td><p>GGplot layer to plot a violing plot</p></td>
      </tr><tr><td>
          <p><code><a href="split_multiallelic.html">split_multiallelic()</a></code> </p>
        </td>
        <td><p>Function to split multiallelic sites into multiple variants</p></td>
      </tr><tr><td>
          <p><code><a href="vcf_to_data_frame.html">vcf_to_data_frame()</a></code> </p>
        </td>
        <td><p>Convert a VCF to a data.frame containing key statistics</p></td>
      </tr><tr><td>
          <p><code><a href="window_list_from_genome.html">window_list_from_genome()</a></code> </p>
        </td>
        <td><p>Creates a data.frame of named tiles from a BSgenome object</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Timon Heide.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer></div>

  


  

  </body></html>

