<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Wrapper function to load arbitrary VCF files. — load_vcf_file • THmisc</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Wrapper function to load arbitrary VCF files. — load_vcf_file" />
<meta property="og:description" content="Wrapper function to load arbitrary VCF files." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">THmisc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper function to load arbitrary VCF files.</h1>
    
    <div class="hidden name"><code>load_vcf_file.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Wrapper function to load arbitrary VCF files.</p>
    </div>

    <pre class="usage"><span class='fu'>load_vcf_file</span><span class='op'>(</span><span class='va'>f</span>, <span class='va'>...</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span>, annot <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>f</th>
      <td><p>File name</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Other optional arguments.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Verbosity flag (Default: TRUE)</p></td>
    </tr>
    <tr>
      <th>annot</th>
      <td><p>Flag indicating if variant annotations should be added (Default: TRUE)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A 'CollapsedVCF' object.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>load_vcf_file</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"platypus.vcf.bgz"</span>, package <span class='op'>=</span> <span class='st'>"THmisc"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span style='font-weight: bold;'>--------------- Loading a VCF file ---------------
#&gt; </span>
#&gt;   File infos:
#&gt;    - Name: '<span style='font-weight: bold;'>platypus.vcf.bgz</span>'
#&gt;    - Created: <span style='font-weight: bold;'>2021-09-02 17:04:18</span> 
#&gt;    - Modified: <span style='font-weight: bold;'>2021-09-02 17:04:16</span> 
#&gt; 
#&gt;   Identifying type... 
#&gt;    =&gt; Type is <span style='color: #00BB00;'>'Platpyus'</span>.
#&gt; 
#&gt;   VCF object contains:
#&gt;    =&gt; N=<span style='font-weight: bold;'>2</span> variants.
#&gt;    =&gt; Of these <span style='font-weight: bold;'>2/2</span> <span style='font-weight: bold;'>(100%)</span> marked passing.
#&gt;    =&gt; On <span style='font-weight: bold;'>1</span> chromosomes
#&gt; 
#&gt;   Guessing reference of a vcf file:</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘BiocGenerics’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:parallel’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
#&gt; <span class='message'>    clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
#&gt; <span class='message'>    parLapplyLB, parRapply, parSapply, parSapplyLB</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    IQR, mad, sd, var, xtabs</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
#&gt; <span class='message'>    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
#&gt; <span class='message'>    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
#&gt; <span class='message'>    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
#&gt; <span class='message'>    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
#&gt; <span class='message'>    union, unique, unsplit, which.max, which.min</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘S4Vectors’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid, I, unname</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘Biostrings’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    strsplit</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘BSgenome.Hsapiens.UCSC.hg19’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:BSgenome.Hsapiens.UCSC.hg38’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Hsapiens</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘BSgenome.Hsapiens.UCSC.hg18’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:BSgenome.Hsapiens.UCSC.hg19’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Hsapiens</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:BSgenome.Hsapiens.UCSC.hg38’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Hsapiens</span></div><div class='output co'>#&gt;    =&gt; Testing 3/3 available options.
#&gt;    =&gt; Found 2/3 perfectly matching.
#&gt;    =&gt; Reference is '<span style='color: #00BB00;'>BSgenome.Hsapiens.UCSC.hg38</span>'.</div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='output co'>#&gt; 
#&gt;   Generating annotations...
#&gt;   
#&gt;  <span style='font-weight: bold;'>--------------------------------------------------</span> </div><div class='output co'>#&gt; class: CollapsedVCF 
#&gt; dim: 2 2 
#&gt; rowRanges(vcf):
#&gt;   GRanges with 6 metadata columns: paramRangeID, ANNOTATION, REF, ALT, QUAL, FILTER
#&gt; info(vcf):
#&gt;   DataFrame with 25 columns: FR, MMLQ, TCR, HP, WE, Source, FS, WS, PP, TR, ...
#&gt; info(header(vcf)):
#&gt;                   Number Type    Description                                   
#&gt;    FR             .      Float   Estimated population frequency of variant     
#&gt;    MMLQ           1      Float   Median minimum base quality for bases aroun...
#&gt;    TCR            1      Integer Total reverse strand coverage at this locus   
#&gt;    HP             1      Integer Homopolymer run length around variant locus   
#&gt;    WE             1      Integer End position of calling window                
#&gt;    Source         .      String  Was this variant suggested by Playtypus, As...
#&gt;    FS             .      Float   Fisher's exact test for strand bias (Phred ...
#&gt;    WS             1      Integer Starting position of calling window           
#&gt;    PP             .      Float   Posterior probability (phred scaled) that t...
#&gt;    TR             .      Integer Total number of reads containing this variant 
#&gt;    NF             .      Integer Total number of forward reads containing th...
#&gt;    TCF            1      Integer Total forward strand coverage at this locus   
#&gt;    NR             .      Integer Total number of reverse reads containing th...
#&gt;    TC             1      Integer Total coverage at this locus                  
#&gt;    END            .      Integer End position of reference call block          
#&gt;    MGOF           .      Integer Worst goodness-of-fit value reported across...
#&gt;    SbPval         .      Float   Binomial P-value for strand bias test         
#&gt;    START          .      Integer Start position of reference call block        
#&gt;    ReadPosRankSum .      Float   Mann-Whitney Rank sum test for difference b...
#&gt;    MQ             .      Float   Root mean square of mapping qualities of re...
#&gt;    QD             1      Float   Variant-quality/read-depth for this variant   
#&gt;    SC             1      String  Genomic sequence 10 bases either side of va...
#&gt;    BRF            1      Float   Fraction of reads around this variant that ...
#&gt;    HapScore       .      Integer Haplotype score measuring the number of hap...
#&gt;    Size           .      Integer Size of reference call block                  
#&gt; geno(vcf):
#&gt;   List of length 7: GT, GQ, GOF, NR, GL, NV, VAF
#&gt; geno(header(vcf)):
#&gt;        Number Type    Description                                              
#&gt;    GT  1      String  Unphased genotypes                                       
#&gt;    GQ  .      Integer Genotype quality as phred score                          
#&gt;    GOF .      Float   Goodness of fit value                                    
#&gt;    NR  .      Integer Number of reads covering variant location in this sample 
#&gt;    GL  .      Float   Genotype log10-likelihoods for AA,AB and BB genotypes,...
#&gt;    NV  .      Integer Number of reads containing variant in this sample        
#&gt;    VAF .      Float   Variant allele frequency                                 </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Timon Heide.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


